### PR TITLE
Minion driver rewrite

### PR DESCRIPTION
**Pull request by babusid**

### CODE REVIEW

File: Drivers/Inc/Minions.h
Comment (babusid) [inline Drivers/Inc/Minions.h:31]: 1) Mark as const
2) Put it in the C file so that it actually compiles

File: Drivers/Src/Minions.c
Comment (babusid) [inline Drivers/Src/Minions.c:8]: no longer need this

File: Drivers/Src/Minions.c
Comment (babusid) [inline Drivers/Src/Minions.c:20]: This isn't generic behavior. This error check should lookup the passed-in pin in the lookup table, and use the returned PinInfo_t struct to determine if the pin is an output. This way, if we add more outputs in the future, we won't have to revisit this.

File: Drivers/Src/Minions.c
Comment (babusid) [inline Drivers/Src/Minions.c:17]:
```suggestion
bool Minion_Read_Input(MinionPin_t pin){
```

File: Drivers/Inc/Minions.h
Comment (babusid) [inline Drivers/Inc/Minions.h:19]: Calling this a Pin Mask doesn't make a ton of sense, I'd call it a `pinNumber` instead

File: Drivers/Src/Minions.c
Comment (babusid) [inline Drivers/Src/Minions.c:27]: Same idea as my earlier comment. This isn't generic behavior. I would suggest having a generic `Minion_Write_Output` function that can take in any pin, and any state, and set it appropriately.

File: Drivers/Src/Minions.c
Comment (IshDeshpa) [inline Drivers/Src/Minions.c:34]: If the direction is wrong, return an error. This requires the creation of an error enum (see [here](https://github.com/lhr-solar/Controls/blob/a4242873c8441336c8aa7f9360020b5b29eadb7c/Drivers/Inc/Display.h#L25-L37)) and an assertion function. Return the error through a pointer parameter and dereference the pointer.

The reasoning behind this is the end user needs to know when a write has failed (also makes it easier to debug)

Actually, we're able to read an output pin, so you may want to return an error here into the parameter (MINION_READ_OUTPUT_ERR or smth) but **_still_** read the input and return it. So returning the error will basically be a warning or a notifier to the user saying "hey you're reading an output but it's not gonna break the car and we're gonna return a valid value anyways". The assertion function can have a switch case that doesn't necessarily go to fault state based off of this.

File: Drivers/Src/Minions.c
Comment (IshDeshpa) [inline Drivers/Src/Minions.c:48]: Same thing here with the error checking.

File: BSP/Inc/BSP_GPIO.h
Comment (IshDeshpa) [inline BSP/Inc/BSP_GPIO.h:56]: param is now a true or false value so the documentation can reflect that

**Review by IshDeshpa — DISMISSED** [review]
Some things to address

File: BSP/Inc/BSP_GPIO.h
Comment (audreygpick) [reply BSP/Inc/BSP_GPIO.h:56]: Nathaniel says you're wrong (and not to change anything in BSP)

File: BSP/STM32F413/Src/BSP_GPIO.c
Comment (babusid) [inline BSP/STM32F413/Src/BSP_GPIO.c:88]: documentation needs to reflect that the boolean true = ON and boolean false = OFF

**Review by babusid — COMMENTED** [review]
lgtm, @millergage, just fix that and we can merge this with the application PR once that's also approved.

File: Drivers/Src/Minions.c
Comment (Cam0Cow) [inline Drivers/Src/Minions.c:44]: This needs a corresponding assertOSError

**Review by Cam0Cow — CHANGES_REQUESTED** [review]

**Review by Cam0Cow — DISMISSED** [review]

**Review by NathanielDelgado — APPROVED** [review]

**Review by Cam0Cow — APPROVED** [review]

**Pull request merged**
